java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java

```diff
@@ -205,46 +220,67 @@ private PreparedStatement prepareDeleteStatement(SinkRow row) {
     public void write(Iterator<SinkRow> rows) {
         while (rows.hasNext()) {
             try (SinkRow row = rows.next()) {
-                PreparedStatement stmt;
                 if (row.getOp() == Data.Op.UPDATE_DELETE) {
                     updateFlag = true;
                     continue;
                 }
-
                 if (config.isUpsertSink()) {
-                    stmt = prepareForUpsert(row);
+                    prepareForUpsert(row);
                 } else {
-                    stmt = prepareForAppendOnly(row);
-                }
-
-                try {
-                    LOG.debug("Executing statement: {}", stmt);
-                    stmt.executeUpdate();
-                    stmt.clearParameters();
-                } catch (SQLException e) {
-                    throw Status.INTERNAL
-                            .withDescription(
-                                    String.format(ERROR_REPORT_TEMPLATE, stmt, e.getMessage()))
-                            .asRuntimeException();
+                    prepareForAppendOnly(row);
                 }
             } catch (Exception e) {
                 throw new RuntimeException(e);
             }
         }
+
+        // execute staging statement in batch after all rows are prepared,
+        // and we need to delete first to prevent accidentally deletion
+        var deleteStmt = stagingStatements.remove(OpType.DELETE);
+        if (deleteStmt != null) {
+            executeStatement(deleteStmt);
+        }
+
+        var upsertStmt = stagingStatements.remove(OpType.UPSERT);
+        if (upsertStmt != null) {
+            executeStatement(upsertStmt);
+        }
+
+        var insertStmt = stagingStatements.remove(OpType.INSERT);
+        if (insertStmt != null) {
+            executeStatement(insertStmt);
+        }
```

Comment:
IIRC we have discussed executing deletes before inserts in sink executor before, but we should do a compaction in the upstream? cc @st1page

Reply:
I am not quite sure whether the execution order is right here, since our e2e test lacks of delete/update workload.

Reply:
@chenzl25 yes, The compaction logic has been implemented in https://github.com/risingwavelabs/risingwave/pull/11070. And I will continue developing that in another PR later.